File: poly/Primes/primes.js
// A very reasonable performance from JavaScript on this one

// TODO: make this work with nodejs

// We did not implement memoization in the thunk. We initially forgot,
// then realized it had no beneficial impact in the Scheme implementation
//
// there is no assert function in javascript
function assert(condition) {
  if (!condition) {
    throw "Assertion failed";
  }
}

function Cell(head, tail){
  this.car = head;
  this.cdr = function(){ return tail; }   // thunk
}

Cell.prototype.first  = function(){ return this.car; }
Cell.prototype.rest   = function(){ return this.cdr(); }
Cell.prototype.take   = function(N){
  assert(N > 0);
  var cell = new Cell(this.first(), null);
  if(N == 1) return cell;
  // defining var func = this.rest will fail for some reason
  var func = this.cdr;  // cannot put 'this' into lambda expression
  cell.cdr = function(){ return func().take(N-1); };
  return cell;
}
Cell.prototype.toString = function(){
  var str = "(";
  var start = true;
  var next = this;
  while(next != null){
    if(!start){
      str += " ";
    }
    else{
      start = false;
    }
    str += next.first();
    next = next.rest();
  }
  return str + ")";
}
Cell.prototype.filter = function(predicate){
  var next = this;
  while(!predicate(next.first())){
    next = next.rest();
  }
  var cell = new Cell(next.first(), null);
  cell.cdr = function(){ return next.rest().filter(predicate); };
  return cell;
}  

Cell.fromTransition = function(init, transition){
  var cell = new Cell(init, null);
  cell.cdr = function(){ 
    return Cell.fromTransition(transition(init), transition); 
  };
  return cell;
}

Cell.sieve = function(input, paramPredicate){
  var x = input.first();
  var cell = new Cell(x, null);
  cell.cdr = function(){
    return Cell.sieve(input.rest().filter(function(n){ 
      return paramPredicate(n,x);
    }), paramPredicate);
  };
  return cell;
}

// 'scriptArgs' works for this interpreter (spidermonkey)
// command line arguments

var numPrimes = Number(process.argv[2]);

from2 = Cell.fromTransition(2, function(n){ return n+1; });
primes = Cell.sieve(from2, function(n,x){ return n % x != 0; });
console.log(primes.take(numPrimes))

File: poly/Primes/Primes.scala
// We did not implement memoization in the thunk. We initially forgot,
// then realized it had no beneficial impact in the Scheme implementation

// quick and dirty implementation of infinite streams
// (which can be finite). We have not defined the empty stream.
class Cell[T](head: T, tail: Cell[T]){
  private val car = head
  private var cdr = () => tail
  def first = car
  def rest  = cdr()

  def filter(p: T => Boolean) : Cell[T] = {
    var next = this;
    while(!p(next.first)){
      next = next.rest
    }
    val succeed = next
    val cell = new Cell(succeed.first,null);
    cell.cdr = () => succeed.rest.filter(p)
    return cell
  }
  
  def take(N: Int) : Cell[T] = {
    assert(N > 0)
    val cell = new Cell(first, null)
    if(N == 1) return cell
    cell.cdr = () => rest.take(N-1)
    return cell 
  } 

  // should rewrite this in functional way
  override def toString  = {
    var str : String = "("
    var next = this;
    while(next != null){
      str += next.first + " "
      next = next.rest
    }
    str + "\b)";
  } 
}

object Cell{
  def fromTransition[T](init: T, f : T => T) : Cell[T] = {
    val cell = new Cell[T](init, null)
    cell.cdr = () => fromTransition(f(init),f)
    return cell
  }

  def sieve[T](input: Cell[T], p: (T,T) => Boolean): Cell[T] = {
    val x = input.first
    val cell = new Cell(x,null)
    cell.cdr = () => sieve(input.rest.filter(n => p(n,x)), p)
    return cell
  }
}

object Primes {
  def main(args: Array[String]){
    val numPrimes = args(0).toInt;
    val from2 = Cell.fromTransition[Int](2, n => n+1)
    val primes = Cell.sieve[Int](from2, (n,x) => n % x != 0)
    println(primes.take(numPrimes))
  }
}
